libnmstate/nm/wired.py

```diff
@@ -0,0 +1,147 @@
+#
+
+import array
+import struct
+import fcntl
+import socket
+
+
+from libnmstate.nm import nmclient
+
+
+ETHTOOL_GSET = 0x00000001  # Get settings
+SIOCETHTOOL = 0x8946
+
+YES = 'yes'
+NO = 'no'
+
+
+def create_setting(iface_state, base_con_profile):
+    mtu = iface_state.get('mtu')
+
+    ethernet = iface_state.get('ethernet', {})
+    speed = ethernet.get('speed')
+    duplex = ethernet.get('duplex')
+    auto_negotiation = ethernet.get('auto-negotiation')
+
+    if not (mtu or speed or duplex or auto_negotiation):
+        return None
+
+    if auto_negotiation == YES:
+        auto_negotiation = True
+
+    if auto_negotiation == NO:
+        auto_negotiation = False
+
+    wired_setting = None
+    if base_con_profile:
+        wired_setting = base_con_profile.get_setting_wired()
+        if wired_setting:
+            wired_setting = wired_setting.duplicate()
+
+    if not wired_setting:
+        wired_setting = nmclient.NM.SettingWired.new()
+
+    if mtu:
+        wired_setting.set_property(nmclient.NM.SETTING_WIRED_MTU, mtu)
+
+    if speed:
+        wired_setting.set_property(nmclient.NM.SETTING_WIRED_SPEED, speed)
+
+    if duplex:
+        wired_setting.set_property(nmclient.NM.SETTING_WIRED_DUPLEX, duplex)
+
+    if auto_negotiation is not None:
+        wired_setting.set_property(nmclient.NM.SETTING_WIRED_AUTO_NEGOTIATE,
+                                   auto_negotiation)
+
+    return wired_setting
+
+
+# based on:
+# https://github.com/rlisagor/pynetlinux/blob/master/pynetlinux/ifconfig.py
+# https://elixir.bootlin.com/linux/v4.19-rc1/source/include/uapi/linux/ethtool.h
+def minimal_ethtool(interface):
+    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    sockfd = sock.fileno()
+
+    ecmd = array.array('B', struct.pack('I39s', ETHTOOL_GSET, b'\x00'*39))
+    ifreq = struct.pack('16sP', interface, ecmd.buffer_info()[0])
+
+    try:
+        fcntl.ioctl(sockfd, SIOCETHTOOL, ifreq)
+        res = ecmd.tostring()
+        speed, duplex, auto = struct.unpack('12xHB3xB24x', res)
+    except IOError:
+        speed, duplex, auto = 65535, 255, 255
+
+    if speed == 65535:
+        speed = 0
+
+    if duplex == 255:
+        duplex = 'unknown'
+    else:
+        duplex = 'full' if bool(duplex) else 'half'
+
+    if auto == 255:
+        auto = None
+    else:
+        auto = bool(auto)
+
+    return dict(speed=speed, duplex=duplex, auto=auto)
+
+
+def get_info(device):
+    """
+    Provides the current active values for a device
+    """
+    info = {}
+
+    iface = device.get_iface()
+    try:
+        info['mtu'] = device.get_mtu()
+    # maybe all device actually have a get_mtu() method
+    except AttributeError:
+        pass
+
+    ethernet = {}
+    try:
+        ethernet['speed'] = device.get_speed()
+    # DeviceWifi, DeviceGeneric, DeviceTun do not have a get_speed() method
+    except AttributeError:
+        pass
+
+    # Duplex state is not available via Network Manager:
```

Comment:
Same, please use the doctext or commit message.
You could also add the BZ as reference.

Reply:
I added the BZ but I do not see how adding this to the commit message or doctext would improve this since the context to the line of code would be lost.

Reply:
How about moving all this to a function, there you can use doctext.